### ISSUE TITLE
Tutorial for ranking modes in CatBoost

### ISSUE THREAD

**Issue by hanky** [catboost/catboost#358]
Hello.

Looks like the current version of CatBoost supports learning to rank. There are some clues about it in the documentation, but I couldn't find any minimal working examples. I wonder which methods should be considered as a baseline approach and what are the prerequisites?

Should we use YetiRank as the training metric and just provide a query id as the Pool group_id parameter? What other CatBoost parameters should be taken into account specifically for a ranking problem? 

Thank you!

**Comment by hanky** [comment]
Thank you for your answer, it was pretty helpful.

Calculated metrics look interesting for me and I have further questions about them.
1. How exactly is PFound calculated? I didn't find any formula in the documentation.
2. How do I calculate NDCG? I didn't find any information about it either and If I pass it as the 'custom_metric' parameter, CatBoost gives me an error: 'NDCG loss is not supported'.

**Comment by annaveronika** [comment]
https://tech.yandex.com/catboost/doc/dg/installation/python-installation-method-build-from-source-linux-macos-docpage/ - here is the instruction for building the library from source

**Comment by Sandy4321** [comment]
yes it would be great to share how to use ranking by using python code

**Comment by uriel-vinetz-rf** [comment]
What does the permutations parameter mean in YetiRank?

**Comment by zpl** [comment]
>We will publish a tutorial for ranking modes at some point

Offering 50$ if tutorial will be published within 1 week (before 23.06.2018)
and extra 50$ if it'll be before monday (18.06.18)

Just some kind of contribute for your great project. thanks.

**Comment by wingrime** [comment]
My YetiRank Code (ListWise Ranking) Example

Fragments from working jupyter notebook. If anyone interested, I can join it with LETOR dataset. When I will have some free time

https://gist.github.com/wingrime/12aa614665c88f4d0254a66efc4ba745

**Comment by zpl** [comment]
@wingrime, thank you for your help. I will count this as reply for my request. so, money sent. =)

**Comment by kizill** [comment]
We added msrank tutorial https://github.com/catboost/catboost/blob/master/catboost/tutorials/advanced_tutorials/catboost_ranking_tutorial.ipynb in 89bdc587fa10086e40

**Comment by saurzcode** [comment]
right path for tutorial = https://github.com/catboost/catboost/blob/master/catboost/tutorials/ranking/ranking_tutorial.ipynb

**Comment by anuragreddygv323** [comment]
looks like the MSrank dataset link is broken

**Comment by iibykov** [comment]
> looks like the MSrank dataset link is broken

Yes.
In lines
`from catboost.datasets import msrank_10k
train_df, test_df = msrank_10k()`

You just need to write
`from catboost.datasets import msrank
train_df, test_df = msrank()`

**Comment by annaveronika** [comment]
In the next release there will be both, msrank and msrank_10k. The msrank one is large, so it takes time to download it. And msrank_10k is the first 10k samples of the whole dataset.

**Comment by ddofer** [comment]
Is there an explanation anywhere on the output of predictions when using ranking? (i.e are "high" scores a high rank or low rank?). There doesn't seem to be any documentation on the ranking models predictions output/format?

**Comment by Sandy4321** [comment]
Good question

**Comment by LyzhinIvan** [comment]
> Is there an explanation anywhere on the output of predictions when using ranking? (i.e are "high" scores a high rank or low rank?). There doesn't seem to be any documentation on the ranking models predictions output/format?

For ranking higher prediction value means higher rank.

**Comment by pro100olga** [comment]
> > looks like the MSrank dataset link is broken
> 
> Yes.
> In lines
> `from catboost.datasets import msrank_10k`
> `train_df, test_df = msrank_10k()`
> 
> You just need to write
> `from catboost.datasets import msrank`
> `train_df, test_df = msrank()`

Looks like the updated version is also broken:

![msrank](https://user-images.githubusercontent.com/9339977/86797356-3d94ec80-c078-11ea-8890-89a3173331c6.PNG)

**Comment by LyzhinIvan** [comment]
> Looks like the updated version is also broken

Which version of catboost do you use?

**Comment by pro100olga** [comment]
> > Looks like the updated version is also broken
> 
> Which version of catboost do you use?

0.11.1

**Comment by LyzhinIvan** [comment]
> > > Looks like the updated version is also broken
> > 
> > 
> > Which version of catboost do you use?
> 
> 0.11.1

It is very old version. msrank() had not been added at 0.11.1 yet. You should update catboost, the latest version is 0.23.2.

**Comment by ddofer** [comment]
Possibly a bug in the ranking documentation, https://catboost.ai/docs/concepts/loss-functions-ranking.html
PairLogitPairwise and PairLogit - "Calculation principles" - displays the exact same formula. Should that be the case? If not, then where's the difference between the 2 methods?

**Comment by LyzhinIvan** [comment]
> Possibly a bug in the ranking documentation, https://catboost.ai/docs/concepts/loss-functions-ranking.html
> PairLogitPairwise and PairLogit - "Calculation principles" - displays the exact same formula. Should that be the case? If not, then where's the difference between the 2 methods?

It is not a bug. These loss function differ in how leaf values are calculated. PairLogit does not use pairs when calculating leaf values, but PairLogitPairwise uses them. You can found details in the paper from PairLogitPairwise documentation.

**Comment by Sandy4321** [comment]
You can found details in the paper from PairLogitPairwise documentation
Can you share link?
